src/NBitcoin/UInt256.cs

```diff
@@ -251,18 +251,32 @@ public uint256(string str)
             if (str.StartsWith("0x", StringComparison.OrdinalIgnoreCase))
                 str = str.Substring(2);
 
-            byte[] bytes = Encoder.DecodeData(str).Reverse().ToArray();
-            if (bytes.Length != WIDTH_BYTE)
-                throw new FormatException("Invalid hex length");
-            this.pn0 = Utils.ToUInt32(bytes, 4 * 0, true);
-            this.pn1 = Utils.ToUInt32(bytes, 4 * 1, true);
-            this.pn2 = Utils.ToUInt32(bytes, 4 * 2, true);
-            this.pn3 = Utils.ToUInt32(bytes, 4 * 3, true);
-            this.pn4 = Utils.ToUInt32(bytes, 4 * 4, true);
-            this.pn5 = Utils.ToUInt32(bytes, 4 * 5, true);
-            this.pn6 = Utils.ToUInt32(bytes, 4 * 6, true);
-            this.pn7 = Utils.ToUInt32(bytes, 4 * 7, true);
-
+            using (var enumerator = Encoder.DecodeDataToUintEnumerable(str).GetEnumerator())
+            {
+                try
+                {
+                    enumerator.MoveNext();
+                    this.pn7 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn6 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn5 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn4 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn3 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn2 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn1 = enumerator.Current;
+                    enumerator.MoveNext();
+                    this.pn0 = enumerator.Current;
+                }
+                catch(InvalidOperationException)
+                {
+                    throw new FormatException("Invalid hex length");
```

Comment:
Nit: this changes functionality as enumerables with more than 8 elements are now possible but there is no longer a check for this.

Reply:
Hmm good spot.

Could do this

` try
                {
                    enumerator.MoveNext();
                    this.pn7 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn6 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn5 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn4 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn3 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn2 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn1 = enumerator.Current;
                    enumerator.MoveNext();
                    this.pn0 = enumerator.Current;
                    if (enumerator.MoveNext())
                    {
                        throw new InvalidOperationException();
                    }
                }
                catch(InvalidOperationException)
                {
                    throw new FormatException("Invalid hex length");
                }`